Add CondVar

brianlheim opened the pull request:
## Purpose and Motivation

A condition variable implementation that intends to replace Condition.
Main features:
- a more intuitive interface
- stronger guarantees about exception safety and blocking/unblocking
  conditions
- predicates are given per-wait instead of per-object
- timeouts
- more closely matches classic condition variable design and what's
  found in other implementations

This grows out of https://github.com/supercollider/supercollider/pull/5368 and specifically my review comment https://github.com/supercollider/supercollider/pull/5368#issuecomment-792300324. Please read that comment if you'd like an understanding of the background for this; it would be helpful but not necessary if you want to review this PR.

Here are questions that I think need to be answered while this is still a draft:
- Is this the right direction generally? Is there anything missing from the interface that should be there, or are any parts of it uncomfortable or confusing?
- Do the preconditions for `timeoutBeats` make sense? Is this code too defensive? If so, how much should we care about exceptions being thrown in a timeout thread or context switch?
- I'm confused by the concept of "threadPlayer". This is not documented anywhere I could find, and it's not entirely clear from the commit message or surrounding code. Why is this concept necessary and why does it affect a relatively low-level synchronization object like CondVar? @telephon this may be a question you are best equipped to answer as it seems you added it originally.
- Do the documentation examples make sense? Or are they too abstract, or place emphasis on the wrong things? Should we consider adding a new document that goes more into detail on how to work with condition variables for those unfamiliar with synchronization/multithreading in general? Or would it be better to instead focus on providing higher-level abstractions to solve more common use cases?

I just want to also add that having prior discussion with @jamshark70 and @telephon on the previous PRs/threads about this topic was a big help for designing this implementation, thanks to you both.

## Types of changes

- New feature

## To-do list

- [x] Get initial feedback from @jamshark70 , @telephon , @iani (I believe you expressed interest at the last dev meeting), and/or others
- [x] Add note in documentation to explain predicate concept ( see https://en.cppreference.com/w/cpp/concepts/predicate)
- [x] Resolve TODOs in code
- [x] Make non-draft

## Future work

After this PR, we make another one to deprecate Condition and replace uses of it in the class library. I think we should also consider some of the functionality from https://github.com/supercollider/supercollider/pull/3895, as this is a pretty skeletal interface and there are features in @scztt 's PR that I think are pretty useful and worth revisiting.

jamshark70 commented:
Hi --

I think the general direction is good. I will have to come back to it in a week or two (my ring finger is currently in a bandage, which affects typing...), and there is (as you said in my older PR) a lot to absorb and understand.

The coding style is quite different from my own -- I expect I'll have some detailed comments on that later. At first glance, `while { predicate.value.not } { this.wait }` for instance is not self-evident -- correct, but I did not "get it" without some effort.

I'll have to wrap up for now, but before that -- `threadPlayer` is for wrapper classes for Routines. `thisThread` is always either a Thread or a Routine. But, a Routine may be under control of another object (Task, TaskProxy, Tdef...).

```supercollider
t = Task {
	"thisThread === t: %\n".postf(thisThread === t);
	"thisThread.threadPlayer === t: %\n".postf(thisThread.threadPlayer === t);
}.play;

thisThread === t: false
thisThread.threadPlayer === t: true
```

The user calls `play` and `stop` on the wrapper object. After `play`, it's the wrapper object (not the Routine) that is in the clock's queue -- therefore, if a thread-in-a-wrapper is waiting for a condition, then it should be the wrapper object that resumes, not only the Routine. So `thisThread` is not sufficient.

Good work here, more than I could have done.

iani commented:
I have trouble cloning or pulling the pull request to examine in depth.

So here are my comments more informally:

It is good to have the examples, and they are clear.

More basic understanding questions - maybe naive - but hopefully useful in
providing understanding for the broader public:

   1. Why do we need to keep an array of waitigThreads? It looks that this
   is for the case that the same condition is called many times from different
   threads?
   2. is there a way to remove a function that has been scheduled for
   execution from a thread by sending it something like
   aThread.remove(aFunction)? This could be handy in other contexts, as for
   example, replacing an array of events scheduled for the future by a
   different array, while the array has already started scheduling some of its
   events. Perhaps this is already  in your code in prRemoveWaitingThread.

/ thread player versions
prSleepThread { |t| waitingThreads = waitingThreads.add(t.threadPlayer) }
prWakeThread { |t| t.clock.sched(0, t) }
prRemoveWaitingThread { |t| ^waitingThreads.remove(t.threadPlayer) }


   1. Last: Why not store the routine that is executing and call .next on
   it when the condition is fulfilled?

OK next I'll look into the methods cited above and comment. Brian, if you
care to say why one cannot use .next in the present context? We have no
access to a routine from the Thread? Or is there some other reason?

Cheers,

Iannis





On Sun, Mar 21, 2021 at 3:11 PM H. James Harkins ***@***.***>
wrote:

> Hi --
>
> I think the general direction is good. I will have to come back to it in a
> week or two (my ring finger is currently in a bandage, which affects
> typing...), and there is (as you said in my older PR) a lot to absorb and
> understand.
>
> The coding style is quite different from my own -- I expect I'll have some
> detailed comments on that later. At first glance, while {
> predicate.value.not } { this.wait } for instance is not self-evident --
> correct, but I did not "get it" without some effort.
>
> I'll have to wrap up for now, but before that -- threadPlayer is for
> wrapper classes for Routines. thisThread is always either a Thread or a
> Routine. But, a Routine may be under control of another object (Task,
> TaskProxy, Tdef...).
>
> t = Task {
> 	"thisThread === t: %\n".postf(thisThread === t);
> 	"thisThread.threadPlayer === t: %\n".postf(thisThread.threadPlayer === t);
> }.play;
> thisThread === t: falsethisThread.threadPlayer === t: true
>
> The user calls play and stop on the wrapper object. After play, it's the
> wrapper object (not the Routine) that is in the clock's queue -- therefore,
> if a thread-in-a-wrapper is waiting for a condition, then it should be the
> wrapper object that resumes, not only the Routine. So thisThread is not
> sufficient.
>
> Good work here, more than I could have done.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/supercollider/supercollider/pull/5394#issuecomment-803519425>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABMNRECX225JIOD2KC2P4TTEWE23ANCNFSM4ZQNGKNQ>
> .
>

brianlheim commented:
@jamshark70 Thanks for looking at it so quickly, and I hope your finger gets better soon!

The explanation about threadPlayer makes sense now. I'll update this PR and also add some comments where I think the code might not be very intuitive.

@iani Thanks for taking a look!

> I have trouble cloning or pulling the pull request to examine in depth.

Do these steps work: https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally

> Why do we need to keep an array of waitigThreads? It looks that this is for the case that the same condition is called many times from different threads?

Correct.

> is there a way to remove a function that has been scheduled for execution from a thread by sending it something like aThread.remove(aFunction)?

I don't understand -- this object schedules thread players, not functions. Is this question about CondVar or about threads in general?

> Why not store the routine that is executing and call .next on it when the condition is fulfilled? [...] if you care to say why one cannot use .next in the present context? We have no access to a routine from the Thread? Or is there some other reason?

I don't understand these questions, sorry. You asked "why not" but I don't see the reason *why* I would do this.

iani commented:
Hello Brian,
thank you for answering. Here my answer to your questions:


is there a way to remove a function that has been scheduled for execution
from a thread by sending it something like aThread.remove(aFunction)?

I don't understand -- this object schedules thread players, not functions.
Is this question about CondVar or about threads in general?


I was referring to a way to undo the timeOut of a condition if it is
fulfilled before the timeOut interval expires. In that case, you would have
to unschedule the function which was scheduled for the timeout - at least
that was what I thought you referred to in your previous comment to Jame's
implementation of timeOut.  Additionally, I am interested in removing from
the scheduler of a thread a function which has been scheduled - in order to
undo a time structure which needs to be replaced but has already started
executing.


Why not store the routine that is executing and call .next on it when the
> condition is fulfilled? [...] if you care to say why one cannot use .next
> in the present context? We have no access to a routine from the Thread? Or
> is there some other reason?
>
I don't understand these questions, sorry. You asked "why not" but I don't
see the reason *why* I would do this.

I was thinking of the possibility of doing

    prWakeThread { |r| r.next }

instead of:

    prWakeThread { |t| t.clock.sched(0, t) }

Perhaps this is a naive understanding question, but it could be helpful to
try and explain in the help file in order to provide a deeper understanding
of how the scheduler and the Threads work together.  I'd be happy to write
up the explanation, but I need some help in how to approach the topic in
order to study it and gain the appropriate understanding.

Iannis



On Sun, Mar 21, 2021 at 5:08 PM Brian Heim ***@***.***> wrote:

> @jamshark70 <https://github.com/jamshark70> Thanks for looking at it so
> quickly, and I hope your finger gets better soon!
>
> The explanation about threadPlayer makes sense now. I'll update this PR
> and also add some comments where I think the code might not be very
> intuitive.
>
> @iani <https://github.com/iani> Thanks for taking a look!
>
> I have trouble cloning or pulling the pull request to examine in depth.
>
> Do these steps work:
> https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally
>
> Why do we need to keep an array of waitigThreads? It looks that this is
> for the case that the same condition is called many times from different
> threads?
>
> Correct.
>
> is there a way to remove a function that has been scheduled for execution
> from a thread by sending it something like aThread.remove(aFunction)?
>
> I don't understand -- this object schedules thread players, not functions.
> Is this question about CondVar or about threads in general?
>
> Why not store the routine that is executing and call .next on it when the
> condition is fulfilled? [...] if you care to say why one cannot use .next
> in the present context? We have no access to a routine from the Thread? Or
> is there some other reason?
>
> I don't understand these questions, sorry. You asked "why not" but I don't
> see the reason *why* I would do this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/supercollider/supercollider/pull/5394#issuecomment-803529515>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABMNRHAY2QQC42AW6BUD43TEWSQPANCNFSM4ZQNGKNQ>
> .
>

iani commented:
PS: the concluding consequence of my last question is there is need to
explain the differences and relationships and dependences between Thread
and Routine. AFAIK Routine depends on Thread to schedule it. But then, why
cannot a Thread call next on a routine? A Thread is responsible for
scheduling the wake up of a Routine which has been stopped by yield or
wait? What is the relationship / difference between .yield, .wait, .sched,
.next?
Thanks and apologies for the insistence, hopefully it is helpful.
Iannis Z.

On Sun, Mar 21, 2021 at 5:25 PM Iannis Zannos ***@***.***> wrote:

> Hello Brian,
> thank you for answering. Here my answer to your questions:
>
>
> is there a way to remove a function that has been scheduled for execution
> from a thread by sending it something like aThread.remove(aFunction)?
>
> I don't understand -- this object schedules thread players, not functions.
> Is this question about CondVar or about threads in general?
>
>
> I was referring to a way to undo the timeOut of a condition if it is
> fulfilled before the timeOut interval expires. In that case, you would have
> to unschedule the function which was scheduled for the timeout - at least
> that was what I thought you referred to in your previous comment to Jame's
> implementation of timeOut.  Additionally, I am interested in removing from
> the scheduler of a thread a function which has been scheduled - in order to
> undo a time structure which needs to be replaced but has already started
> executing.
>
>
> Why not store the routine that is executing and call .next on it when the
>> condition is fulfilled? [...] if you care to say why one cannot use .next
>> in the present context? We have no access to a routine from the Thread? Or
>> is there some other reason?
>>
> I don't understand these questions, sorry. You asked "why not" but I don't
> see the reason *why* I would do this.
>
> I was thinking of the possibility of doing
>
>     prWakeThread { |r| r.next }
>
> instead of:
>
>     prWakeThread { |t| t.clock.sched(0, t) }
>
> Perhaps this is a naive understanding question, but it could be helpful to
> try and explain in the help file in order to provide a deeper understanding
> of how the scheduler and the Threads work together.  I'd be happy to write
> up the explanation, but I need some help in how to approach the topic in
> order to study it and gain the appropriate understanding.
>
> Iannis
>
>
>
> On Sun, Mar 21, 2021 at 5:08 PM Brian Heim ***@***.***>
> wrote:
>
>> @jamshark70 <https://github.com/jamshark70> Thanks for looking at it so
>> quickly, and I hope your finger gets better soon!
>>
>> The explanation about threadPlayer makes sense now. I'll update this PR
>> and also add some comments where I think the code might not be very
>> intuitive.
>>
>> @iani <https://github.com/iani> Thanks for taking a look!
>>
>> I have trouble cloning or pulling the pull request to examine in depth.
>>
>> Do these steps work:
>> https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally
>>
>> Why do we need to keep an array of waitigThreads? It looks that this is
>> for the case that the same condition is called many times from different
>> threads?
>>
>> Correct.
>>
>> is there a way to remove a function that has been scheduled for execution
>> from a thread by sending it something like aThread.remove(aFunction)?
>>
>> I don't understand -- this object schedules thread players, not
>> functions. Is this question about CondVar or about threads in general?
>>
>> Why not store the routine that is executing and call .next on it when the
>> condition is fulfilled? [...] if you care to say why one cannot use .next
>> in the present context? We have no access to a routine from the Thread? Or
>> is there some other reason?
>>
>> I don't understand these questions, sorry. You asked "why not" but I
>> don't see the reason *why* I would do this.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/supercollider/supercollider/pull/5394#issuecomment-803529515>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AABMNRHAY2QQC42AW6BUD43TEWSQPANCNFSM4ZQNGKNQ>
>> .
>>
>

brianlheim commented:
@iani can you please make sure to use quote characters (>) to clearly show when you are quoting an earlier statement, as I am doing? Your previous comment put everything on the same level and it's confusing to read.

> I was referring to a way to undo the timeOut of a condition if it is
> fulfilled before the timeOut interval expires. In that case, you would have
> to unschedule the function which was scheduled for the timeout - at least
> that was what I thought you referred to in your previous comment to Jame's
> implementation of timeOut.

I am [already doing this](https://github.com/supercollider/supercollider/blob/3ad7f1e3aa0379b1f6ec1ec244f456644d00d1dd/SCClassLibrary/Common/Core/CondVar.sc#L52).

> Additionally, I am interested in removing from
> the scheduler of a thread a function which has been scheduled - in order to
> undo a time structure which needs to be replaced but has already started
> executing.

This is unrelated to the PR, so I'm not going to respond here. You're free to ask this somewhere else, though.

> I was thinking of the possibility of doing
> 
>     prWakeThread { |r| r.next }
> 
> instead of:
> 
>     prWakeThread { |t| t.clock.sched(0, t) }
> 
> Perhaps this is a naive understanding question, but it could be helpful to
> try and explain in the help file in order to provide a deeper understanding
> of how the scheduler and the Threads work together.  I'd be happy to write
> up the explanation, but I need some help in how to approach the topic in
> order to study it and gain the appropriate understanding.

The purpose of CondVar.schelp is to document the purpose and usage of CondVar. These strike me as implementation details, so I disagree that explaining them there is appropriate. I do think we could document the threadPlayer concept better and then include a small note in CondVar.schelp which links to that, but I don't want to do that in this PR.

Briefly, as I understand it, scheduling on thread players handles this at the right level of abstraction because of what James said above -- we want to be scheduling routines in their wrapped context, and not the routines directly. James or Julian may be able to provide more context for this if it's still unclear.

>  PS: the concluding consequence of my last question is there is need to
> explain the differences and relationships and dependences between Thread
> and Routine. AFAIK Routine depends on Thread to schedule it. But then, why
> cannot a Thread call next on a routine? A Thread is responsible for
> scheduling the wake up of a Routine which has been stopped by yield or
> wait? What is the relationship / difference between .yield, .wait, .sched,
> .next?

@iani , these questions are *very far* outside the scope of this PR. It's okay to ask a few clarifying questions to get the right context for reviewing, but any more than that and the conversation should really take place somewhere more appropriate for learning. So I would encourage you to start a thread on scsynth.org for it, and/or read some of the related code and documentation.

iani commented:
ok I stand corrected.
Thanks for answering about removing a scheduled thread.
I'll take the other questions elsewhere after reviewing and reformulating
them.
Regarding the PR, I'll look into the code and wait to see other comments.

Best,

Iannis

On Sun, Mar 21, 2021 at 6:12 PM Brian Heim ***@***.***> wrote:

> @iani <https://github.com/iani> can you please make sure to use quote
> characters (>) to clearly show when you are quoting an earlier statement,
> as I am doing? Your previous comment put everything on the same level and
> it's confusing to read.
>
> I was referring to a way to undo the timeOut of a condition if it is
> fulfilled before the timeOut interval expires. In that case, you would have
> to unschedule the function which was scheduled for the timeout - at least
> that was what I thought you referred to in your previous comment to Jame's
> implementation of timeOut.
>
> I am already doing this
> <https://github.com/supercollider/supercollider/blob/3ad7f1e3aa0379b1f6ec1ec244f456644d00d1dd/SCClassLibrary/Common/Core/CondVar.sc#L52>
> .
>
> Additionally, I am interested in removing from
> the scheduler of a thread a function which has been scheduled - in order to
> undo a time structure which needs to be replaced but has already started
> executing.
>
> This is unrelated to the PR, so I'm not going to respond here. You're free
> to ask this somewhere else, though.
>
> I was thinking of the possibility of doing
>
> prWakeThread { |r| r.next }
>
> instead of:
>
> prWakeThread { |t| t.clock.sched(0, t) }
>
> Perhaps this is a naive understanding question, but it could be helpful to
> try and explain in the help file in order to provide a deeper understanding
> of how the scheduler and the Threads work together. I'd be happy to write
> up the explanation, but I need some help in how to approach the topic in
> order to study it and gain the appropriate understanding.
>
> The purpose of CondVar.schelp is to document the purpose and usage of
> CondVar. These strike me as implementation details, so I disagree that
> explaining them there is appropriate. I do think we could document the
> threadPlayer concept better and then include a small note in CondVar.schelp
> which links to that, but I don't want to do that in this PR.
>
> Briefly, as I understand it, scheduling on thread players handles this at
> the right level of abstraction because of what James said above -- we want
> to be scheduling routines in their wrapped context, and not the routines
> directly. James or Julian may be able to provide more context for this if
> it's still unclear.
>
> PS: the concluding consequence of my last question is there is need to
> explain the differences and relationships and dependences between Thread
> and Routine. AFAIK Routine depends on Thread to schedule it. But then, why
> cannot a Thread call next on a routine? A Thread is responsible for
> scheduling the wake up of a Routine which has been stopped by yield or
> wait? What is the relationship / difference between .yield, .wait, .sched,
> .next?
>
> @iani <https://github.com/iani> , these questions are *very far* outside
> the scope of this PR. It's okay to ask a few clarifying questions to get
> the right context for reviewing, but any more than that and the
> conversation should really take place somewhere more appropriate for
> learning. So I would encourage you to start a thread on scsynth.org for
> it, and/or read some of the related code and documentation.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/supercollider/supercollider/pull/5394#issuecomment-803538376>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABMNRAI662AA5Q7NSE5OVDTEW2BLANCNFSM4ZQNGKNQ>
> .
>

telephon commented on SCClassLibrary/Common/Core/CondVar.sc:
```diff
@@ -0,0 +1,105 @@
+
+	*new { ^super.newCopyArgs(Array.new, Array.new) }
+
+	wait { |predicate|
```
This will loop "full speed" until the predicate returns true. Is that absolutely required? It seems risky and one needs to know what one is doing.

Isn't `signal` supposed to make such a while loop unnecessary`

brianlheim replied:
I think you might be misunderstanding -- this isn't a busy loop; each `wait` call inside the loop blocks until woken via signalling, and only when the thread is woken from a signal does it check the predicate again. If it's true, we exit the loop and return from the wait; otherwise, we go back to waiting.

telephon replied:
ah yes, now I can spot the trick, the inner wait doesn't pass the predicate so that it'll sleep. Intricate :)

brianlheim replied:
both you and James had trouble understanding this, so i'm planning to add a second private function which both branches will call down to, and do the same for waitFor. I think that will make it easier to understand

muellmusik commented:
Just catching up on this. No specific comments at this time, but just wanted to thank @brianlheim for all the thoughtful and detailed work and investigation that’s gone into this!

telephon commented on SCClassLibrary/Common/Core/CondVar.sc:
```diff
@@ -0,0 +1,105 @@
+	prRemoveWaitingThread { |t| ^waitingThreads.remove(t.threadPlayer) }
+
+	// Precondition checks for `waitFor` timeout. May throw.
+	prHandleTimeoutPreconditions { |n|
```
"handle preconsitions" is a bit general for what it does. But I would suggest to introduce an interface for the respective classes, because it is generally useful and sematically unambiguous in all other use cases, I think. Then name `asTimeoutValue` could also be something more general like `asDeltaTime`? Here I assume for simplicity that an Object that responds to `asInteger` will also respond to `asFloat`.

```
+ SimpleNumber {
	asTimeoutValue {
		if(this.isNaN) { Error("timeout value should be a number").throw };
		if(this === inf ) { Error("timeout value should be finite").throw };
		if(this < 0 ) { Error("timeout value should be positive").throw };
		^this
	}
	
}

+ Object {
	
	asTimeoutValue {
		^this.asFloat.asTimeoutValue
	}
}

```

Then this method isn't needed, and above you'd write instead of:

`timeoutBeats = this.prHandleTimeoutPreconditions(timeoutBeats);`

`timeoutBeats = timeoutBeats.asTimeoutValue`

brianlheim replied:
I won't do that here, but you are welcome to try it out in a future PR if you find more use cases for this concept.

brianlheim replied:
at this point, I am really just looking for feedback on whether the restrictions on the type and value make sense. to me, they seem intuitive (except for maybe inf and -inf), and i think it's important to not allow an exception to propagate out of the spawned timeout thread, but this still feels like a rough edge in the design to me and I'd be open to other approaches

telephon replied:
Yes, this is a good procedure, you can do it this way. In the final version, these conversions should be replaced by something less particular (I don't remember any other case where we have something like this), but that can wait for a bit later if you prefer.

telephon commented on SCClassLibrary/Common/Core/CondVar.sc:
```diff
@@ -0,0 +1,105 @@
+		}
+	}
+
+	signalOne {
```
in similar cases in sclang, we have pairings of `something` and `somethingAll` – so just `signal` might be more idiomatic.

brianlheim replied:
I don't mind doing this. The old interface used `unhang` and `notify` to mean "unhang all" and "notify all", so I thought it would be better to be explicit here in case people are used to the old interface.

telephon replied:
I see, that makes sense. Still the word "one" immediately rises the question for me: "which one?" Should it be `signalFirst` or `signalNext`?

brianlheim replied:
There are a lot of condition variable implementations out there we can draw from for inspiration. I'd encourage you to go take a look at some and then see if you still feel the same way about these names. The ones I looked at while reviewing James's PR were the implementations from C++ (which is the one closest to the design I used here), WinAPI, pthreads, Java, D, Rust, Python, and Go.

telephon replied:
Ok, good! I've checked some, and I see that notify_one seems to be a model for c++ oriented implementations. In Ruby and Java, the combination seems to be  `signal` and `broadcast`, which sounds good. But I see your point, so why not keep it.

telephon commented on HelpSource/Classes/CondVar.schelp:
```diff
@@ -0,0 +1,415 @@
+class::CondVar
```
It would be good to explain why this is a "variable", but unlike e.g. a `FlowVar` it doesn't hold a value.

brianlheim replied:
yes, good point, I'm not quite sure where the name "condition variable" comes from as they aren't any more variable-like to me than other synchronization objects like locks, mutexes, semaphores, etc.

brianlheim replied:
would this wording work for you? "Although it has "variable" in the name, a condition variable doesn't wrap a variable like, for example, FlowVar does."

telephon replied:
Yes, that would be good! While popular, it still is somewhat a misnomer.

brianlheim replied:
done!

telephon reviewed (commented):
Thanks! For now just some initial comments.

brianlheim commented:
@jamshark70 @telephon thanks again for the initial feedback. Since the loops in wait and waitFor seem to be pretty confusing, I've refactored them a bit and also added comments to explain some of the trickier parts of the implementation. Hope that helps!

jamshark70 commented on SCClassLibrary/Common/Core/CondVar.sc:
```diff
@@ -0,0 +1,126 @@
+		} {
+			// By taking thisThread.beats here, we wait a little longer than we
+			// should because of the time it takes to get from the start of `waitFor`
+			// to this line, but that's alright since precise timeouts aren't an
```
Just a spot comment for now -- `thisThread.beats` is logical time, not physical time. Between requesting the wait/timeout and calculating the ending beads, there is no `yield`, so logical time advances by 0. So it won't "wait a little longer" -- measured in terms of logical time, it should be exact. (Measured by physical time, there would be some deviation, but that's true of a simple `dur.wait` as well.)

brianlheim replied:
thanks for pointing this out! I forgot SC's clocks have this concept. I've updated my code to reflect this. I thought through the implementation again with this new information and I don't think it requires any code changes. I'd appreciate if you could suggest if/how I should note this in the documentation.

this now opens up the possibility that a predicate could modify thisThread.beats and muck with the waitFor timeout logic. I don't know if this is a case worth worrying about, I'd also be happy to just stick a note in the documentation to the effect of "do this at your own risk". what do you think?

jamshark70 replied:
I agree -- if anyone starts messing around with the thread's or the clock's beats, there is no longer any guarantee of time values advancing in a natural way. (I'd take this as a natural enough assumption that it might not even need to be documented.)

Hoping to get a chance to look at this more carefully over the weekend.

jamshark70 commented on SCClassLibrary/Common/Core/CondVar.sc:
```diff
@@ -0,0 +1,116 @@
+			this.prWait
+		} {
+			// If the predicate is already true, we return immediately. If not,
+			// we block until the predicate is true after being signalled.
```
For "bears of little brain" (in this case, including myself), it would be helpful to note in the comment that the condition may be signaled many times before the predicate becomes true -- hence, the loop is necessary. I honestly didn't understand that at first reading, and I expect others will struggle with it.

jamshark70 commented on SCClassLibrary/Common/Core/CondVar.sc:
```diff
@@ -0,0 +1,116 @@
+
+			didNotTimeout = false;
+
+			// wokenThread must be non-nil. If the thread woke before this timeout fired,
```
"If the condition was signalled before this timeout fired, this routine will be stopped (below, after prWait) and we won't get here" -- I didn't get *how* the thread would awake (it should never wake without a signal, right?).

brianlheim replied:
>  (it should never wake without a signal, right?).

correct, you could still reschedule it yourself I guess, but I don't think that would have any effect beyond what signal does. I haven't thought through that very carefully, so I could be wrong. in my view, not something to protect against, but I can think it through more if you feel differently.

jamshark70 commented on SCClassLibrary/Common/Core/CondVar.sc:
```diff
@@ -0,0 +1,116 @@
+		waitingThreads = Array.new;
+	}
+
+	shallowCopy { this.shouldNotImplement(thisMethod) }
```
Maybe worth noting in a comment that `copy` dispatches to Object:copy --> CondVar:shallowCopy?

jamshark70 reviewed (changes requested):
Carefully done, and usage is clear. I like it. Comments are minor.

I have to admit that I didn't read through the unit tests thoroughly. At a superficial glance, it looks like it's done right (covers *a lot* of territory, and the methodology appears to be sound). I guess what I would have to do to evaluate the tests is to break CondVar and make sure that they fail, but I can't quite manage that today.

brianlheim commented on HelpSource/Classes/CondVar.schelp:
```diff
@@ -0,0 +1,414 @@
+	// certain that the message is not empty. No other threads run in the time between that check and the return from
+	// `wait`, so the only way for the predicate to become false is if it depends on state external to the interpreter,
+	// like the system time.
+	c.wait({ message.notEmpty });
```
I don't have a strong opinion about trailing closure syntax, so whatever you think is easiest for people to understand is fine with me.

I've reflowed the code and comments in the examples to 70 characters, but I'd like to note that if you want this to be standard for contributed patches, it isn't very helpful to bring it up in PRs without documenting that somewhere.

brianlheim replied:
for what it's worth, i also agree that 60 or 70 character lines look best, there just isn't a convention for this so i did what was easiest under my default vim configuration.

i got the impression from past discussions that many non-expert users don't find trailing closure syntax intuitive, which is why i avoided it here. but again, i'll do whatever you think is best. i have a slight preference for keeping it consistent -- i.e. if we change all the waits we should change waitFor, too.

telephon replied:
We have both `wait` and `waitFor`, and (as james notes) `c.waitFor(5) { … }` really falls apart (this is not the case with `if`, because we are so used to it). So I'd tend towards keeping both parentheses (`c.wait({ … })`) in this case. But no strong opinion ...

jamshark70 replied:
Let me preface this by saying I'm OK to leave it alone, on grounds that `waitFor(timeOut) { predicate }` looks clumsy and that examples should be syntactically consistent.

`if(cond) { true_branch }` and `do(number_or_array) { body }` don't fall apart, I think, because the concepts of a conditional branch or loop body resemble control structures that are familiar from other languages. A predicate here isn't quite one of those.

At the same time, I'll also note that literally just four hours ago, Daniel Mayer [commented](https://scsynth.org/t/some-observations/3464/37) on the forum, "Coming from Lisp, the frequent mix of different kinds of brackets looked odd to me. That can be overcome, though, with indentation and dropping unnecessary brackets (e.g. `do { }` instead of `do({ })`)." It isn't the first time I've heard that remark. I would not be at all surprised if we start seeing examples in the wild of `cond.wait { ... }` (and I'll probably do it this way in my own code).

> I've reflowed the code and comments in the examples to 70 characters, but I'd like to note that if you want this to be standard for contributed patches, it isn't very helpful to bring it up in PRs without documenting that somewhere.

I'm just reacting to what I see:

![wrap](https://user-images.githubusercontent.com/318301/112738210-bc7ddc80-8f9b-11eb-84bf-e4596cd9f91a.png)

Yeah, I think we should discourage this in helpfile examples. (FWIW, I don't recall this ever coming up before -- there's no trend that I should have noticed and proposed a style guide change. That is, we're noticing it *now*.)

In any case, with a few search engine queries and poking at a couple of pages on the wiki, I can't even *find* the official style guide. So... not today.

brianlheim replied:
https://github.com/supercollider/supercollider/wiki/Code-style-guidelines

brianlheim replied:
> Let me preface this by saying I'm OK to leave it alone, on grounds that waitFor(timeOut) { predicate } looks clumsy and that examples should be syntactically consistent.

OK, in this case I will just leave it the way it is.

brianlheim commented on HelpSource/Classes/CondVar.schelp:
```diff
@@ -0,0 +1,414 @@
+other words, the threads which were signalled do not run again until the current thread suspends execution, for example
+by calling code::yield:: or by reaching the end of a playing Routine.
+
+In other implementations, you may see that condition variables also use a mutex for synchronization. This is typically
```
i've changed it to "in other languages"

brianlheim commented on HelpSource/Classes/CondVar.schelp:
```diff
@@ -0,0 +1,414 @@
+
+Internally, CondVar is little more than a queue of threads which are waiting for the condition to become true. When
+signaling, you have the choice of waking only the next waiting thread on the queue, if one exists (code::signalOne::),
+or waking all waiting threads in the queue (code::signalAll::). Note that signalling is not a yielding operation. In
```
thanks, I've used this wording

brianlheim commented on HelpSource/Classes/CondVar.schelp:
```diff
@@ -0,0 +1,414 @@
+equivalent to code::while { predicate.value.not } { cond.wait }::. In other words, the thread only blocks if the
+predicate is false, otherwise it blocks and only resumes once the thread has been signalled and the predicate is true.
+
+You must not be on the main thread when using this method.
```
thanks, changed it

brianlheim commented:
> Carefully done, and usage is clear. I like it. Comments are minor.

Thanks James, I appreciate it!

> I have to admit that I didn't read through the unit tests thoroughly. At a superficial glance, it looks like it's done right (covers a lot of territory, and the methodology appears to be sound). I guess what I would have to do to evaluate the tests is to break CondVar and make sure that they fail, but I can't quite manage that today.

I did TDD for some so I know they'll break if the implementation changes slightly, but I'd really appreciate if you could take the time to go over them carefully, because they touch on some important edge cases. Anyway, thanks for the review!

jamshark70 commented:
> I did TDD for some so I know they'll break if the implementation changes slightly, but I'd really appreciate if you could take the time to go over them carefully, because they touch on some important edge cases.

That'll be a couple weeks... I just can't put a lot of time into it in the next few days.

brianlheim commented:
> That'll be a couple weeks... I just can't put a lot of time into it in the next few days.

OK! @telephon would you have time to review this before then perhaps?

brianlheim commented:
> In any case, with a few search engine queries and poking at a couple of pages on the wiki, I can't even find the official style guide.

What made it difficult to find in your experience? Does the wiki just have too many pages? I'm asking because I already know where most of the pages are on the wiki and what they're named, so it's difficult for me to imagine how it looks to someone who doesn't visit it frequently.

Btw, if you want to discuss that style change further, open a new ticket please.

telephon reviewed (changes requested):
This looks very good – so all this is just final touches. Some of it is merely a suggestion. I've run all examples and tests.

I didn't fake any failures for the tests, however. But you can never catch everything anyhow …

brianlheim commented on HelpSource/Classes/CondVar.schelp:
```diff
@@ -0,0 +1,438 @@
+		task = counter;
+		counter = counter + 1;
+		queue = queue.add(task);
+		postf("Producer created a task: %. Current queue size: %\n",
```
thanks for catching this, i copy-pasted this from the below code carelessly!

brianlheim commented on SCClassLibrary/Common/Core/CondVar.sc:
```diff
@@ -0,0 +1,122 @@
+		} {
+			endingBeats = thisThread.beats + timeoutBeats;
+
+			// If the predicate is already true, we return true immediately.
```
in my mind this applies to the behavior of the whole loop -- i'm trying to say that if the predicate is already true, we skip the entire loop body and go straight to the return statement. but maybe that's obvious already, i'll just delete this line i think.

brianlheim commented:
thanks for the review @telephon ! i'll make these edits tomorrow

jamshark70 commented:
> What made it difficult to find in your experience? 

First I did a Bing search for "supercollider style guide," with no relevant results.

So then I searched for "supercollider wiki," OK, but then... is this a "contributing" or "development" topic? The links under "development" didn't seem relevant... but above that there was "Contributing." At the bottom of this page, yesterday I had overlooked "Coding guidelines" because I was looking for the keyword "style" -- today I noticed it, but "Sorry this page does not exist =( "

The lack of a search feature for the wiki is a usability impairment, I think.

> Btw, if you want to discuss that style change further, open a new ticket please.

I think I will have to decline to do that. As you know, lately I have been trying to extricate myself from PR obligations (to the point, in fact, of preparing to close my active PRs). I consented to review this PR as a courtesy, but I didn't consent to assume the lead on side topics resulting from review comments.

brianlheim commented:
> The lack of a search feature for the wiki is a usability impairment, I think.

fyi -- the github search bar, when searching within a repository, also searches wiki pages. definitely not the most discoverable feature, but it is possible at least.

what I meant to ask is is there anything concrete we could do to make those guidelines more easily findable?

> I didn't consent to assume the lead on side topics resulting from review comments.

ok, no worries!

brianlheim commented:
@telephon i've addressed your comments, this is ready for another review. :)

brianlheim commented on HelpSource/Classes/CondVar.schelp:
```diff
@@ -0,0 +1,438 @@
+// and the condition state, `message`.
+fork {
+	waitThreadDelay.wait;
+	// It might be tempting to use `if` here instead of `while`,
```
this comment felt a bit like editorializing to me so I cut it down to a single sentence about how using while is typical, and moved it up to the spot where i explain the difference between the with and without predicate approaches

brianlheim commented on HelpSource/Classes/CondVar.schelp:
```diff
@@ -0,0 +1,438 @@
+
+fork {
+	waitThreadDelay.wait;
+	// As a predicate, we pass a function indicating the condition we
```
i ended up removing most of this comment and turning it into a small note in wait.

telephon commented:
One last thing, concerning the naming of `signalAll` and `signalOne`. I haven't yet understood how to decide which one to use. This would be good to explain in the helpfile, best under these respective methods.

There is one comment in the code:
```
			// Even though there might be multiple consumers waiting
			// on this condition, only one of them can possibly make
			// use of the new task we just created.  So we use
			// signalOne instead of signalAll.
```
Unless I really start guessing, I don't understand it. What is the task in question, why can only one make use of it, and why calling signalAll would make more tasks make use of it? 

Is there a common case where one need not worry about the difference? If there is, perhaps one could just add an explanation when it needs special care. And also name the respective method simply `signal`, and not worry, because the more complicated case (be it `signalAll` or `signalOne`) would be known.

telephon commented:
My last comment wasn't intended to hold up the process. Let's briefly think about how to proceed – I'd be entirely fine with it as it is. The "one last thing" was really just something I would enjoy to think through, but not a demand.

dyfer commented:
@telephon @brianlheim what's the status of this PR?

brianlheim commented:
Sorry I couldn't finish this up. If someone takes it over, please think through carefully how you will deprecate Condition; I made a small amount of progress on this and it's not trivial.

<3

joshpar commented:
Rather than closing - is there value in pulling in CondVar, using it for our Unit Tests, and keeping Condition around for the time being? @brianlheim - I'll trust your judgement on that.

brianlheim commented:
You can do whatever you'd like with it, I'm no longer maintaining this branch or PR.

joshpar commented:
@dyfer or @telephon - thoughts? It seemed like we were good to go with this at some layer.

telephon commented:
Yes, definitely. I didn't know we were about to deprecate Condition, that is something for the further future.

dyfer commented:
@telephon so what needs to be done to finish this up? I wasn't following the details of the implementation closely.

I am also in favor of not deprecating Condition for the time being.

telephon commented:
I'd be fine with merging it as it is and perhaps adjusting in a separate step. I had a question, but maybe I am just a bit dumb …

joshpar commented:
> Carefully done, and usage is clear. I like it. Comments are minor.
> 
> I have to admit that I didn't read through the unit tests thoroughly. At a superficial glance, it looks like it's done right (covers _a lot_ of territory, and the methodology appears to be sound). I guess what I would have to do to evaluate the tests is to break CondVar and make sure that they fail, but I can't quite manage that today.

Any concerns with merging @jamshark70 ? @brianlheim will not be making further changes, and I believe this satisfies the need in our unit tests for this to be pulled in.

patrickdupuis commented:
Is there a way to use `CondVar.wait` or `CondVar.waitFor` so that a blocked thread continues soon after the predicate becomes `true`? I'm thinking about the case of unit tests. Often, you want wait for a asynchronous command to finish (with a timeout) and move on quickly. From what I understand, in order to have `CondVar.waitFor` return early when its predicate becomes `true`, you need to periodically signal it. Am I right?

dyfer commented:
Closing in favor of #5436 in order to remove a debug commit. We plan to commit these changes soon. Thanks everyone!

avdrd commented:
As this appears to have been the design discussion central for CondVar, I'll leave a brief note here that while CondVar, despite being just a pretty standard implementation of a Mesa aka signal-and-continue monitor is way better than the original Condition implementation, it seems clear from the discussion above that the dev who wrote it was not fully familiar with scalng internals, as they were still unsure what threadPlayer does etc., at one point at least. 

Something that was actually missed in the above (too long) discussions was that in sclang you can definitely have the equivalent of spurious wake-ups as far as CondVar is concerned, meaning that a routine may be resumed from a wait without having been signaled, e.g. simply some other thread calling `next` on it directly. In the present CondVar implementation this results in the routine adding multiple entries of itself (or more precisely of its threadPlayer) to the waitingThreads queue, so it will run multiple times later e.g. on signalAll.

Some say that a routine being inovoked by means other than a signal is a monitor protocol violation, so that scenario need not be supported by CondVar, but coding around this limitation was pretty hard for me, so I've rolled my own CondVar class that handles those spurious wake-ups automagically, i.e. doesn't result in duplicate entries for the same thread in waitingThreads,. Alas, I did it in the most convenient way for me, which meant dropping the FIFO activation guarantee that CondVar provides (modulo duplicate activations), so I'm not offering my code as a bug fix.
